src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs

```diff
@@ -608,12 +582,27 @@ private void CheckDisposed()
                         {
                             ResetTimeoutTimer(timeoutTimer);
 
-                            // We have data, process it
-                            var (close, exception) = await ProcessMessagesAsync(buffer, connectionState);
+                            Log.ProcessingMessage(_logger, buffer.Length);
+
+                            var close = false;
+
+                            while (_protocol.TryParseMessage(ref buffer, connectionState, out var message))
+                            {
+                                Exception exception;
+
+                                // We have data, process it
+                                (close, exception) = await ProcessMessagesAsync(message, connectionState);
+                                if (close)
+                                {
+                                    // Closing because we got a close frame, possibly with an error in it.
+                                    connectionState.CloseException = exception;
+                                    break;
+                                }
+                            }
+
+                            // If we're closing stop everything
                             if (close)
                             {
-                                // Closing because we got a close frame, possibly with an error in it.
-                                connectionState.CloseException = exception;
                                 break;
```

Comment:
Two-level `break`.... 

![https://i.giphy.com/media/KvWUKOYR8phja/giphy.webp](https://i.giphy.com/media/KvWUKOYR8phja/giphy.webp)

Reply:
I could add a break label 😄

Reply:
Yeah, I don't like that either. Oh well, I'll just deal with it.